### PR TITLE
[BEAM-9926] Programming guide - Fix typos and add clearer Python text

### PR DESCRIPTION
**Pull request by epicfaace**
Fix some typos and make docs clearer for Python in the Programming Guide. R: @melap 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: website/www/site/content/en/documentation/programming-guide.md [hunk 1]
```diff
@@ -683,18 +687,18 @@ that for you. Your `@ProcessElement` method should accept a parameter tagged wit
...
 {{< paragraph class="language-py" >}}
 Inside your `DoFn` subclass, you'll write a method `process` where you provide
 the actual processing logic. You don't need to manually extract the elements
 from the input collection; the Beam SDKs handle that for you. Your `process`
-method should accept an object of type `element`. This is the input element and
-output is emitted by using `yield` or `return` statement inside `process`
-method.
+method should accept an argument `element`, which is the input element, emit
+output value(s) by using `yield` statements. You can also use a `return`
+statement if you are sure the method will return only a single output value.
```
Comment (pabloem) [inline website/www/site/content/en/documentation/programming-guide.md:701]: The reason that `yield` is used is that `process` should return an _iterable_. `yield` is the way to pass individual elements from an iterable (does that make sense?)

Return statements can be used, but they must return an iterable - for example a list. So, for example, these functions are equivalent:

```
def process(self, elm):
  yield elm
```

and
```
def process(self, elm):
  return [elm]
```

Comment (pabloem) [reply website/www/site/content/en/documentation/programming-guide.md:701]: So, perhaps I would write something like...

```
Your `process` method should accept an argument `element`, which is the input element, 
and return an iterable with its output values. You can accomplish this by emitting individual 
elements with `yield` statements. You can also use a `return` statement with an iterable,
like a list or a generator.
```

I am not married to the phrasing - whichever phrasing you choose is fine. But I think it's important to mention the fact that an iterable is expected.

Thanks @epicfaace

**Comment by pabloem** [comment]
This looks great. Just added one comment.

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
r: @rosetn

File: website/www/site/content/en/documentation/programming-guide.md (hunk 1 above)
Comment (epicfaace) [reply website/www/site/content/en/documentation/programming-guide.md]: That sounds good!

**Comment by pabloem** [comment]
retest this please

**Comment by pabloem** [comment]
this LGTM. I'll wait for @rosetn to add her comments as well

File: website/www/site/content/en/documentation/programming-guide.md [hunk 2]
```diff
@@ -527,7 +530,7 @@ The graph of this pipeline looks like the following:
 *Figure 1: A linear pipeline with three sequential transforms.*
 
 However, note that a transform *does not consume or otherwise alter* the input
-collection--remember that a `PCollection` is immutable by definition. This means
+collection - remember that a `PCollection` is immutable by definition. This means
```
Comment (rosetn) [inline website/www/site/content/en/documentation/programming-guide.md:533]: Em dash with no spaces between words is the correct usage here: —

File: website/www/site/content/en/documentation/programming-guide.md [hunk 3]
```diff
@@ -1842,6 +1847,9 @@ transform's intermediate data changes type multiple times.
 {{< github_sample "/apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/snippets.py" pipeline_monitoring_composite >}}
 {{< /highlight >}}
 
+Note that because `Count` is itself a composite transform,
```
Comment (rosetn) [inline website/www/site/content/en/documentation/programming-guide.md:1850]: I'd either format this as a note (see other notes on this page) or remove "Note that"

Comment (pabloem) [reply website/www/site/content/en/documentation/programming-guide.md]: done

File: website/www/site/content/en/documentation/programming-guide.md (hunk 2 above)
Comment (pabloem) [reply website/www/site/content/en/documentation/programming-guide.md]: done

**Comment by pabloem** [comment]
thanks @epicfaace -I'll merge after tests pass

**Comment by pabloem** [comment]
thanks @epicfaace @rosetn !

**Pull request merged**